src/duct.py

```diff
@@ -164,62 +166,165 @@ def create_and_parse_args():
+        choices=["all", "system-summary", "processes-samples"],
+        help="Record system-summary, processes-samples, or all",
+    )
     return parser.parse_args()
 
 
+class TeeStream:
+    """TeeStream simultaneously streams to standard output (stdout) and a specified file."""
+
+    def __init__(self, file_path):
+        self.file = open(file_path, "w")
+        (
+            self.listener_fd,
+            self.writer_fd,
+        ) = os.openpty()  # Use pseudo-terminal to simulate terminal behavior
+
+    def fileno(self):
+        """Return the file descriptor to be used by subprocess as stdout/stderr."""
+        return self.listener_fd
+
+    def start(self):
+        """Start a thread to read from the main_fd and write to stdout and the file."""
+        thread = threading.Thread(target=self._redirect_output, daemon=True)
+        thread.start()
+
+    def _redirect_output(self):
+        with os.fdopen(self.listener_fd, "rb", buffering=0) as stream:
+            while True:
+                try:
+                    data = stream.read(1024)
+                except OSError as e:
+                    if e.errno == errno.EIO:  # The file has been closed
+                        break
+                    else:
+                        raise
+                if not data:  # Still open, but no new data to write
+                    break
+                sys.stdout.buffer.write(data)
+                sys.stdout.buffer.flush()
+                self.file.write(
+                    data.decode("utf-8", "replace")
+                )  # Handling decoding errors
+                self.file.flush()
+
+    def close(self):
+        """Close the slave fd and the file when done."""
+        os.close(self.listener_fd)
+        self.file.close()
+
+
+def monitor_process(
+    stdout, stderr, report, process, report_interval, sample_interval, output_prefix
+):
+    while True:
+        if process.poll() is not None:  # the passthrough command has finished
+            if isinstance(stdout, TeeStream):
+                stdout.close()
+            if isinstance(stderr, TeeStream):
+                stderr.close()
+            break
+
+        elapsed_time = time.time() - report.start_time
+        report.collect_sample()
+        if elapsed_time >= (report.number + 1) * report_interval:
+            aggregated = report.aggregate_samples()
+            for pid, pinfo in aggregated.items():
+                with open(
+                    f"{output_prefix}/{pid}_resource_usage.json", "a"
+                ) as resource_statistics_log:
+                    pinfo["elapsed_time"] = elapsed_time
+                    resource_statistics_log.write(json.dumps(aggregated))
+            report.number += 1
+        time.sleep(sample_interval)
+
+
+def prepare_outputs(capture_outputs, outputs, output_prefix):
+    if capture_outputs in ["all", "stdout"] and outputs in ["all", "stdout"]:
+        stdout = TeeStream(f"{output_prefix}/stdout.txt")
```

Comment:
let's not breed folders unless asked to:
- if `output_prefix` ends with `/` -- ensure that folder exists
- here/somewhere before if to do once do `output_prefix.format()` giving those variables to format
- just do `f"{output_prefix_formatted}stdout"` (and stderr below) so it would be for prefix to contain `_` separator. (I have reservations against .txt  since there is no guarantee that it is really a .txt -- may be it was some `dd` from a binary file, or may be it is json inside?... and probably no benefit from having extension)